e2e-tests/vanilla/common/BrowserWindow.py

```diff
@@ -0,0 +1,261 @@
+        def on_draw_event(_, de):
+            nonlocal timeout
+
+            GLib.source_remove(timeout)
+            timeout = GLib.timeout_add(draw_timeout / 2, on_timeout)
+            return False
+
+        signal_id = self.web_view.connect("draw", on_draw_event)
+        loop.run()
+        self.web_view.disconnect(signal_id)
+
+        overlay.destroy()
+        self.web_view.grab_focus()
+
+    def send_key(self, event_type, key):
+        default_seat = Gdk.Display.get_default().get_default_seat()
+        event = Gdk.Event.new(event_type)
+        event.set_device(default_seat.get_keyboard())
+        event.set_source_device(default_seat.get_keyboard())
+        event.window = self.web_view.get_window()
+        event.send_event = True
+        event.keyval = key
+
+        loop = GLib.MainLoop()
+
+        def on_event(_, event):
+            if event.type == event_type and event.keyval == key:
+                loop.quit()
+            return False
+
+        self.web_view.connect("event", on_event)
+        event.put()
+        loop.run()
+
+    def send_key_tap(self, key):
+        self.send_key(Gdk.EventType.KEY_PRESS, key)
+        self.send_key(Gdk.EventType.KEY_RELEASE, key)
+
+    def send_key_taps(self, key_taps):
+        for kt in key_taps:
+            self.send_key_tap(kt)
+
+
+class TOTPGenerator:
+    def __init__(self, otp_secret):
+        self.otp_secret = otp_secret
+
+    def generate_totp(self) -> str:
+        intervals_number = int(time.time()) // 30
+        token = str(self.__get_hotp_token(self.otp_secret, intervals_no=intervals_number))
+        # Pad token up to 6 digits
+        while len(token) != 6:
+            token += '0'
+
+        return token
+
+    def __get_hotp_token(self, secret, intervals_no):
+        key = base64.b32decode(secret, True)
+        msg = struct.pack(">Q", intervals_no)
+        h = hmac.new(key, msg, hashlib.sha1).digest()
+        o = o = h[19] & 15
+        h = (struct.unpack(">I", h[o:o + 4])[0] & 0x7fffffff) % 1000000
+        return h
+
+
+def ascii_string_to_key_events(string):
+    if len(string) != len(string.encode()):
+        raise TypeError(f"{string} is not an ascii string")
+    return [ord(ch) for ch in list(string)]
+
+
+def wait_for_successful_text(webview, selector="h1.p-heading--4", retries=10, delay=1.0):
+    js = f'''
+        (function() {{
+            var el = document.querySelector("{selector}");
+            return el ? el.innerText : "";
+        }})();
+    '''
+
+    for _ in range(retries):
+        result_holder = {"text": ""}
+
+        def on_js_finished(webview, result, user_data):
+            js_result = webview.evaluate_javascript_finish(result)
+            result_holder["text"] = js_result.to_string()
+
+        webview.evaluate_javascript(js, -1, None, None, None, on_js_finished, None)
+
+        while Gtk.events_pending():
+            Gtk.main_iteration()
+
+        if result_holder["text"]:
+            return result_holder["text"]
+
```

Comment:
This is not something we should do, as it may break many things.

Instead the code should be looking something like:

```python
    js_result = None
    loop = GLib.MainLoop()

    def on_js_finished(webview, result):
        nonlocal js_result

        try:
            js_result = webview.evaluate_javascript_finish(result)
        except Exception as e: # TODO: filter exception...
            print(f"{e}")
            # We want to raise it too?
            # raise e
        loop.quit()

    webview.evaluate_javascript(js, -1, None, None, None, on_js_finished, None)
    loop.run()

    if js_result and js_result.to_string():
        return js_result.to_string()

    wait_for_timeout(delay)
```

Using: https://github.com/ubuntu/authd-oidc-brokers/pull/659#discussion_r2368294909